### PR TITLE
[CNI] zap logger for platform package

### PR DESCRIPTION
**Pull request by paulyufan2**
**Reason for Change**:

This PR is to migrate platform package logger to zap logger

Remove log from GetLastRebootTime() and log it at caller

**Issue Fixed**:

**Requirements**:

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] relevant PR labels added

**Notes**:

### CODE REVIEW

File: platform/os_windows.go [hunk 1]
```diff
@@ -132,11 +142,11 @@ func SetOutboundSNAT(subnet string) error {
 // This will be called only when reboot is detected - This is windows specific
 func ClearNetworkConfiguration() (bool, error) {
 	jsonStore := CNIRuntimePath + "azure-vnet.json"
-	log.Printf("Deleting the json store %s", jsonStore)
+	logger.Info("Deleting the json", zap.String("store", jsonStore))
```
[the remark this answers is not in the document]
Comment (paulyufan2) [reply platform/os_windows.go:145]: This ClearNetworkConfiguration() is used by network package (manager.go)

File: platform/os_windows.go [hunk 2]
```diff
@@ -156,7 +166,7 @@ func ExecutePowershellCommand(command string) (string, error) {
 		return "", fmt.Errorf("Failed to find powershell executable")
 	}
 
-	log.Printf("[Azure-Utils] %s", command)
+	logger.Info("[Azure-Utils]", zap.String("command", command))
```
[the remark this answers is not in the document]
Comment (paulyufan2) [reply platform/os_windows.go:169]: This is used by network package (endpoint_windows.go) and platform mellanox.go

File: network/manager.go [hunk 3]
```diff
@@ -155,6 +155,9 @@ func (nm *networkManager) restore(isRehydrationRequired bool) error {
 		modTime, err := nm.store.GetModificationTime()
 		if err == nil {
 			rebootTime, err := platform.GetLastRebootTime()
+			if err != nil {
+				logger.Info("Failed to query last reboot time", zap.Error(err))
```
Comment (vipul-21) [inline network/manager.go:159]: This should be `logger.Error`

File: cni/network/plugin/main.go [hunk 4]
```diff
@@ -185,6 +185,8 @@ func rootExecute() error {
 		upTime, err = platform.GetLastRebootTime()
 		if err == nil {
 			cniReport.VMUptime = upTime.Format("2006-01-02 15:04:05")
+		} else {
+			logger.Info("Failed to query last reboot time", zap.Error(err))
```
Comment (vipul-21) [inline cni/network/plugin/main.go:189]: This would be `logger.Error`

File: platform/os_linux.go [hunk 5]
```diff
@@ -64,15 +64,13 @@ func GetLastRebootTime() (time.Time, error) {
 	// Query last reboot time.
 	out, err := exec.Command("uptime", "-s").Output()
 	if err != nil {
-		log.Printf("Failed to query uptime, err:%v", err)
```
Comment (vipul-21) [inline platform/os_linux.go:67]: Nice idea to handle from the parent ! :D

File: platform/os_windows.go [hunk 6]
```diff
@@ -13,13 +13,17 @@ import (
 	"syscall"
```
Comment (vipul-21) [inline platform/os_windows.go:13]: We are changing only the functions used by CNI right ? I see `log` and `zaplogger` both in this file

Comment (paulyufan2) [reply platform/os_windows.go:13]: yes that's correct, we only change functions called by CNI, other packages are still using legacy log

**Review by vipul-21 — APPROVED** [review]

File: platform/os_windows.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (tamilmani1989) [reply platform/os_windows.go:163]: is this function used by cns?

[the remark this answers is not in the document]
Comment (tamilmani1989) [reply platform/os_windows.go:163]: looks its called from cns , we have to write wrapper as like store package.. pass logger as argument when initializing platform client .. MonitorAndSetMellanoxRegKeyPriorityVLANTag -> eventually calls ExecutePowershellCmd

**Review by tamilmani1989 — CHANGES_REQUESTED** [review]

File: platform/os_windows.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (paulyufan2) [reply platform/os_windows.go]: ExecutePowershellCommand() and executePowershellCommand() are only used for CNS

File: platform/os_windows.go [hunk 7]
```diff
@@ -13,13 +13,17 @@ import (
...
+	zaplog "github.com/Azure/azure-container-networking/cni/log"
 	"github.com/Azure/azure-container-networking/log"
 	"github.com/Azure/azure-container-networking/platform/windows/adapter"
 	"github.com/Azure/azure-container-networking/platform/windows/adapter/mellanox"
 	"github.com/pkg/errors"
+	"go.uber.org/zap"
 	"golang.org/x/sys/windows"
 )
 
+var logger = zaplog.CNILogger.With(zap.String("component", "platform-windows"))
```
Comment (tamilmani1989) [inline platform/os_windows.go:25]: how `zaplog.CNILogger` initialized for cns? should be moved as argument to platform client...

Comment (vipul-21) [reply platform/os_windows.go:25]: @paulyufan2  Do we need this if we are passing the zap logger

Comment (paulyufan2) [reply platform/os_windows.go:25]: Yes. In the platform only funcs in os_windows.go are zap logged.

var logger = zaplog.CNILogger.With(zap.String("component", "platform-windows"))
logger is used in ClearNetworkConfiguration() only in this file, so we need to initialize this in this file.
I did pass logger as a parameter in ExecutePowerShellCommand()

File: cni/network/plugin/main.go [hunk 8]
```diff
@@ -182,9 +182,12 @@ func rootExecute() error {
 		cniReport.GetReport(pluginName, version, ipamQueryURL)
 
 		var upTime time.Time
-		upTime, err = platform.GetLastRebootTime()
+		p := platform.NewExecClient(logger)
+		upTime, err = p.GetLastRebootTime()
 		if err == nil {
 			cniReport.VMUptime = upTime.Format("2006-01-02 15:04:05")
+		} else {
```
Comment (tamilmani1989) [inline cni/network/plugin/main.go:189]: can we log this inside GetLastRebootTime so that not each caller need to log it

File: cns/restserver/util.go [hunk 9]
```diff
@@ -539,8 +539,11 @@ func (service *HTTPRestService) restoreNetworkState() error {
 
 	if err == nil {
 		logger.Printf("[Azure CNS] Store timestamp is %v.", modTime)
-
-		rebootTime, err := platform.GetLastRebootTime()
+		p := platform.NewExecClient(nil)
+		rebootTime, err := p.GetLastRebootTime()
+		if err != nil {
+			logger.Errorf("Failed to query last reboot time due to %v", err)
+		}
```
Comment (tamilmani1989) [inline cns/restserver/util.go:546]: same as above comment

File: iptables/iptables.go [hunk 10]
```diff
@@ -88,7 +88,7 @@ type IPTableEntry struct {
 func RunCmd(version, params string) error {
 	var cmd string
 
-	p := platform.NewExecClient()
+	p := platform.NewExecClient(nil)
```
Comment (tamilmani1989) [inline iptables/iptables.go:91]: can we initialize logger for iptable package.. iptable package is used by cni

File: network/endpoint_windows.go [hunk 11]
```diff
@@ -183,7 +184,9 @@ func (nw *network) addIPv6NeighborEntryForGateway(epInfo *EndpointInfo) error {
 		// run powershell cmd to set neighbor entry for gw ip to 12-34-56-78-9a-bc
 		cmd := fmt.Sprintf("New-NetNeighbor -IPAddress %s -InterfaceAlias \"%s (%s)\" -LinkLayerAddress \"%s\"",
 			nw.Subnets[1].Gateway.String(), containerIfNamePrefix, epInfo.Id, defaultGwMac)
-		if out, err = platform.ExecutePowershellCommand(cmd); err != nil {
+
+		platformLogger := zaplog.CNILogger.With(zap.String("component", "platform-windows"))
+		if out, err = platform.ExecutePowershellCommand(cmd, platformLogger); err != nil {
```
Comment (tamilmani1989) [inline network/endpoint_windows.go:189]: should update this

File: network/manager.go [hunk 12]
```diff
@@ -152,14 +152,18 @@ func (nm *networkManager) restore(isRehydrationRequired bool) error {
...
 
 	if isRehydrationRequired {
+		var p = platform.NewExecClient(logger)
 		modTime, err := nm.store.GetModificationTime()
 		if err == nil {
-			rebootTime, err := platform.GetLastRebootTime()
+			rebootTime, err := p.GetLastRebootTime()
+			if err != nil {
+				logger.Error("Failed to query last reboot time", zap.Error(err))
+			}
```
Comment (tamilmani1989) [inline network/manager.go:161]: should be moved inside GetLastRebootTime

File: platform/os_windows.go [hunk 13]
```diff
@@ -130,33 +143,37 @@ func SetOutboundSNAT(subnet string) error {
...
+		p.logger.Info("Error deleting the json", zap.String("store", jsonStore))
 		return true, err
 	}
 
 	return true, nil
 }
 
 func KillProcessByName(processName string) {
-	p := NewExecClient()
+	p := NewExecClient(nil)
```
Comment (tamilmani1989) [inline platform/os_windows.go:160]: isn't this function used by CNI..if so i would suggest to pass execClient as receiver to this api

File: platform/os_windows.go [hunk 14]
```diff
@@ -130,33 +143,37 @@ func SetOutboundSNAT(subnet string) error {
...
 	cmd := fmt.Sprintf("taskkill /IM %v /F", processName)
 	p.ExecuteCommand(cmd)
 }
 
 // ExecutePowershellCommand executes powershell command
-func ExecutePowershellCommand(command string) (string, error) {
+func ExecutePowershellCommand(command string, logger *zap.Logger) (string, error) {
```
Comment (tamilmani1989) [inline platform/os_windows.go:166]: execClient should be receiver for this api

File: telemetry/telemetry_windows.go [hunk 15]
```diff
@@ -37,7 +38,7 @@ func (report *CNIReport) GetSystemDetails() {
 }
 
 func (report *CNIReport) GetOSDetails() {
-	p := platform.NewExecClient()
+	p := platform.NewExecClient(nil)
```
Comment (tamilmani1989) [inline telemetry/telemetry_windows.go:41]: should initialize or pass logger to report Structure

**Review by tamilmani1989 — CHANGES_REQUESTED** [review]

File: platform/os_windows.go (hunk 13 above)
Comment (paulyufan2) [reply platform/os_windows.go:160]: It's used by telemetry package and does not log anything here, so we can leave it here

File: cni/network/plugin/main.go [hunk 16]
```diff
@@ -182,9 +182,12 @@ func rootExecute() error {
...
+			logger.Error("Failed to query last reboot time", zap.Error(err))
```
Comment (vipul-21) [inline cni/network/plugin/main.go:190]: This should be logged in platform package right ?

File: cni/telemetry/service/telemetrymain.go [hunk 17]
```diff
@@ -127,7 +127,7 @@ func main() {
 
 	logger.Info("Config path", zap.String("path", configPath))
 
-	config, err = telemetry.ReadConfigFile(configPath)
+	config, err = tb.ReadConfigFile(configPath)
```
Comment (vipul-21) [inline cni/telemetry/service/telemetrymain.go:130]: I don't think this is correct. Did we create a new telemetry buffer to use the ReadConfigFile ?

File: platform/os.go [hunk 18]
```diff
@@ -4,8 +4,9 @@ import (
 	"bufio"
 	"fmt"
 	"io"
-	"log"
 	"os"
+
+	"github.com/Azure/azure-container-networking/log"
```
Comment (vipul-21) [inline platform/os.go:9]: Is this change required for zaplogger ?

Comment (paulyufan2) [reply platform/os.go:9]: No, it's a bug here that it should use legacy log instead of go-buildin log "import log"; and this package is used by CNS only, so this does not need to be zap logged.

File: cni/network/plugin/main.go (hunk 16 above)
Comment (paulyufan2) [reply cni/network/plugin/main.go]: solved

File: network/manager.go [hunk 19]
```diff
@@ -152,14 +152,15 @@ func (nm *networkManager) restore(isRehydrationRequired bool) error {
 	}
 
 	if isRehydrationRequired {
+		p := platform.NewExecClient(logger)
```
Comment (tamilmani1989) [inline network/manager.go:155]: can't we use ExecClient from networkmanager?

File: platform/os_windows.go (hunk 13 above)
Comment (tamilmani1989) [reply platform/os_windows.go:161]: telemetry package is used by CNI..so the logger should be passed from there

File: telemetry/telemetry_windows.go (hunk 15 above)
Comment (tamilmani1989) [reply telemetry/telemetry_windows.go:41]: instead of nil, it should receive logger from CNIReport?

**Review by tamilmani1989 — CHANGES_REQUESTED** [review]

File: telemetry/telemetrybuffer.go [hunk 20]
```diff
@@ -331,25 +338,17 @@ func (tb *TelemetryBuffer) StartTelemetryService(path string, args []string) err
...
 	config := TelemetryConfig{}
 
 	b, err := os.ReadFile(filePath)
 	if err != nil {
-		if tb.logger != nil {
-			tb.logger.Error("Failed to read telemetry config", zap.Error(err))
-		} else {
-			log.Logf("[Telemetry] Failed to read telemetry config: %v", err)
-		}
+		log.Logf("[Telemetry] Failed to read telemetry config: %v", err)
```
Comment (tamilmani1989) [inline telemetry/telemetrybuffer.go]: no need to log here.. let the caller log it

File: telemetry/telemetrybuffer.go [hunk 21]
```diff
@@ -331,25 +338,17 @@ func (tb *TelemetryBuffer) StartTelemetryService(path string, args []string) err
...
 		return config, err
 	}
 
 	if err = json.Unmarshal(b, &config); err != nil {
-		if tb.logger != nil {
-			tb.logger.Error("unmarshal failed with", zap.Error(err))
-		} else {
-			log.Logf("[Telemetry] unmarshal failed with %v", err)
-		}
+		log.Logf("[Telemetry] unmarshal failed with %v", err)
```
Comment (tamilmani1989) [inline telemetry/telemetrybuffer.go]: same as above

**Review by tamilmani1989 — CHANGES_REQUESTED** [review]

File: network/endpoint_windows.go [hunk 22]
```diff
@@ -183,7 +184,10 @@ func (nw *network) addIPv6NeighborEntryForGateway(epInfo *EndpointInfo) error {
 		// run powershell cmd to set neighbor entry for gw ip to 12-34-56-78-9a-bc
 		cmd := fmt.Sprintf("New-NetNeighbor -IPAddress %s -InterfaceAlias \"%s (%s)\" -LinkLayerAddress \"%s\"",
 			nw.Subnets[1].Gateway.String(), containerIfNamePrefix, epInfo.Id, defaultGwMac)
-		if out, err = platform.ExecutePowershellCommand(cmd); err != nil {
+
+		platformLogger := zaplog.CNILogger.With(zap.String("component", "cni-platform-windows"))
```
[the remark this answers is not in the document]
Comment (tamilmani1989) [reply network/endpoint_windows.go:188]: what about this?

[the remark this answers is not in the document]
Comment (paulyufan2) [reply network/endpoint_windows.go]: done

File: ipam/manager.go [hunk 23]
```diff
@@ -134,7 +134,11 @@ func (am *addressManager) restore(rehydrateIpamInfoOnReboot bool) error {
 		// Check if the VM is rebooted.
 		modTime, err := am.store.GetModificationTime()
 		if err == nil {
-			rebootTime, err := platform.GetLastRebootTime()
+			p := platform.NewExecClient(nil)
+			rebootTime, err := p.GetLastRebootTime()
+			if err != nil {
```
Comment (vipul-21) [inline ipam/manager.go:139]: Do we need this check as it was not there before and we already log the error in Getlastrebootime func.?

File: telemetry/telemetrybuffer.go [hunk 24]
```diff
@@ -299,17 +303,36 @@ func WaitForTelemetrySocket(maxAttempt int, waitTimeInMillisecs time.Duration) {
 }
 
 // StartTelemetryService - Kills if any telemetry service runs and start new telemetry service
-func StartTelemetryService(path string, args []string) error {
-	platform.KillProcessByName(TelemetryServiceProcessName)
+func (tb *TelemetryBuffer) StartTelemetryService(path string, args []string) error {
+	err := platform.NewExecClient(tb.logger).KillProcessByName(TelemetryServiceProcessName)
```
Comment (vipul-21) [inline telemetry/telemetrybuffer.go:307]: Shouldn't this be added as exec client to the telemetry buffer ?

File: network/endpoint_windows.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (vipul-21) [reply network/endpoint_windows.go]: I think what TM mentioned here is that we already have execclient in the `newEndpointImpl` function. It is being passed but not used currently. https://github.com/Azure/azure-container-networking/blob/v1.4.39.2/network/endpoint_windows.go#L66. We can pass the same  platform execClient and not create a new one.

File: telemetry/telemetrybuffer.go (hunk 24 above)
Comment (paulyufan2) [reply telemetry/telemetrybuffer.go:307]: KillProcessByName is added to execClient interface

**Review by tamilmani1989 — DISMISSED** [review]

File: network/network_linux.go [hunk 25]
```diff
@@ -337,7 +337,7 @@ func applyDnsConfig(extIf *externalInterface, ifName string) error {
 		setDnsList string
 		err        error
 	)
-	p := platform.NewExecClient()
+	p := platform.NewExecClient(logger)
```
Comment (vipul-21) [inline network/network_linux.go:340]: this should be plcclient of `nm` ?

File: network/network_linux.go [hunk 26]
```diff
@@ -246,7 +246,7 @@ func isGreaterOrEqaulUbuntuVersion(versionToMatch int) bool {
 func readDnsInfo(ifName string) (DNSInfo, error) {
 	var dnsInfo DNSInfo
 
-	p := platform.NewExecClient()
+	p := platform.NewExecClient(logger)
```
Comment (vipul-21) [inline network/network_linux.go:249]: The same exec client we have for nm manager should be used here right ? similar to windows one?

File: network/network_linux.go [hunk 27]
```diff
@@ -443,7 +443,7 @@ func (nm *networkManager) connectExternalInterface(extIf *externalInterface, nwI
 	isGreaterOrEqualUbuntu17 := isGreaterOrEqaulUbuntuVersion(ubuntuVersion17)
 	isSystemdResolvedActive := false
 	if isGreaterOrEqualUbuntu17 {
-		p := platform.NewExecClient()
+		p := platform.NewExecClient(logger)
```
Comment (vipul-21) [inline network/network_linux.go:446]: Same here

File: network/network_linux.go (hunk 25 above)
Comment (vipul-21) [reply network/network_linux.go:339]: Is this resolved ?

File: platform/os_linux.go [hunk 28]
```diff
@@ -115,14 +128,18 @@ func SetOutboundSNAT(subnet string) error {
...
-	p := NewExecClient()
+func (p *execClient) ExecutePowershellCommand(_ string) (string, error) {
+	return "", nil
+}
+
+func (p *execClient) KillProcessByName(processName string) error {
+	newExecClient := NewExecClient(p.logger)
 	cmd := fmt.Sprintf("pkill -f %v", processName)
-	_, err := p.ExecuteCommand(cmd)
+	_, err := newExecClient.ExecuteCommand(cmd)
```
Comment (vipul-21) [inline platform/os_linux.go:142]: Why this change ? Shouldn't this be `p.ExecuteCommand` ?

File: platform/os_linux.go [hunk 29]
```diff
@@ -53,34 +54,46 @@ func GetOSInfo() string {
 }
 
 func GetProcessSupport() error {
```
Comment (vipul-21) [inline platform/os_linux.go:56]: I assume the function that are suing their own client are a part of CNS or other packages only and not cni ?

File: platform/os_linux.go (hunk 28 above)
Comment (paulyufan2) [reply platform/os_linux.go:142]: line 140 newExecClient := NewExecClient(p.logger)

File: platform/os_linux.go (hunk 29 above)
Comment (paulyufan2) [reply platform/os_linux.go:56]: This one is not used anywhere actually, not sure whether we should remove it

File: platform/os_linux.go [hunk 30]
```diff
@@ -115,12 +128,15 @@ func SetOutboundSNAT(subnet string) error {
...
 // ClearNetworkConfiguration clears the azure-vnet.json contents.
 // This will be called only when reboot is detected - This is windows specific
-func ClearNetworkConfiguration() (bool, error) {
+func (p *execClient) ClearNetworkConfiguration() (bool, error) {
 	return false, nil
 }
 
-func KillProcessByName(processName string) error {
-	p := NewExecClient()
+func (p *execClient) ExecutePowershellCommand(_ string) (string, error) {
```
Comment (vipul-21) [inline platform/os_linux.go:135]: @tamilmani1989 This is a bit odd as this is linux file. Would be nice to change the name altogether to `ExecShellCommand` or something.

**Review by vipul-21 — APPROVED** [review]

**Review by tamilmani1989 — APPROVED** [review]
